### PR TITLE
chore: require acking v0.4 upgrade instructions via env var

### LINKED ISSUE #5603: Check that shutdown happened in lockstep
We could have a env var `FM_I_READ_THE_RELEASE_LOG` that can be set to `0.4.0` and we can check whenever we require lockstep upgrades.

### PR DESCRIPTION
**Pull request by dpc**
This will require all end users to acknowledge that they are aware of upgrading in the lockstep requirement.

Fix #5603

### CODE REVIEW

**Review by joschisan — APPROVED** [review]

**Comment by dpc** [comment]
@douglaz Do you think this is OK w.r.t. to infra automation etc.? I think it shouldn't be a problem, but would love a second opinion.

**Comment by douglaz** [comment]
Looks ok to me. But for people running on third party services like Clovyr it will be pretty much impossible to safely upgrade, I guess.

**Comment by dpc** [comment]
Ping @Kodylow for opinion.

**Comment by dpc** [comment]
> But for people running on third party services like Clovyr it will be pretty much impossible to safely upgrade, I guess.

@douglaz You mean due to the lockstep upgrade requirement, right? Or is setting env var an extra problem?

File: fedimintd/src/fedimintd.rs [hunk 1]
```diff
@@ -464,6 +465,37 @@ impl Fedimintd {
...
+    // The actual code-path we want end users to take
+    if std::env::var(FM_REL_NOTES_ACK_ENV).unwrap_or_default() == FM_REL_NOTES_CUR_VAL {
+        return Ok(());
+    }
+
+    eprintln!(
+        "This version of fedimintd has some criticallly important requirements you should be aware of."
+    );
+    eprintln!("Not following them will likely result in a consensus failure or other problems.");
+    eprintln!("To ensure you have read them, you must set specific environment variable to a specific value.");
```
Comment (Kodylow) [inline fedimintd/src/fedimintd.rs:492]: change to "Read release notes for instructions on how to perform the upgrade"

**Comment by Kodylow** [comment]
ACK. yes this doesn't work with clovyr, no one is running mainnet clovyrs besides me though so that's fine.

File: fedimintd/src/fedimintd.rs (hunk 1 above)
Comment (dpc) [reply fedimintd/src/fedimintd.rs:492]: @Kodylow What's the reasoning behind the change.

One thing I like to highlight is setting this env (and thus this message) is displayed irrespective of the fact of setting up or upgrading the Federation. So "how to perform the upgrade" might be confusing for the user that is not doing any upgrade.

File: fedimintd/src/fedimintd.rs [hunk 2]
```diff
@@ -464,6 +465,37 @@ impl Fedimintd {
     }
 }
 
+fn check_release_notes_ack() -> anyhow::Result<()> {
+    const VERSION: &str = "v0.4";
+    const FM_SKIP_REL_NOTES_ACK_ENV: &str = "FM_SKIP_REL_NOTES_ACK";
+    const FM_REL_NOTES_ACK_ENV: &str = "FM_REL_NOTES_ACK";
+    // the suffix here is to prevent people getting to smart and trying to
```
Comment (m1sterc001guy) [inline fedimintd/src/fedimintd.rs:472]: nit: "too" :)

**Review by m1sterc001guy — APPROVED** [review]
LGTM

**Comment by fedimint-backports** [comment]
Successfully created backport PR for `releases/v0.4`:
- #5616

**Pull request merged**